GSTextureCache: Palette management rework. Huge performance boost in ZoE2 with 8-bit textures enabled.

AlessandroVetere opened the pull request:
This follows the 4th point of #2310.
The idea was to rework the current palette texture management to improve performances with games that uses multiple palettes for the same data texture.

Currently the situation goes more or less like that for textures with palette:

- 32 bits texture enabled (i.e. palettes are applied by the CPU):
    
    - On `Source` creation:
        - Clut array of 256 `uint32` elements is allocated and cleared (also for textures without palette),
        - Current clut content from `GSRenderer` is copied to the new buffer
    - On `Source` cache lookup:
        - Current `GSRenderer` clut content is compared against the copy in the `Source`.
    - If palette associated to the texture changes:
        - A new `Source` object with a different clut buffer is computed.
    - On `Source` deletion:
        - Free clut buffer.

- 8 bit texture enabled (i.e. palettes are managed by the GPU, so palette texture upload is required):
    - On `Source` creation:
        - Clut array of 256 `uint32` elements is allocated and cleared,
        - Current clut content from `GSRenderer` is copied to the new buffer,
        - A new Palette `GSTexture` object is created and current clut content is uploaded to the GPU.
    - On `Source` cache lookup:
        - If first lookup then:
            - Current clut content is uploaded to the GPU,
        - else:
            - Update the `Source` clut buffer from `GSRenderer` clut, and if the content mismatches, upload the new content to the same `GSTexture`.
    - If palette associated to the texture changes:
        - As said above, the new content is uploaded to the GPU
    - On `Source` deletion:
        - Free clut buffer,
        - Recycle palette `GSTexture` object.

The new proposed management works like that instead:

- 32 bits texture enabled:
    - On `Source` creation:
        - Clut array of 16 or 256 `uint32` elements is allocated (and not cleared) only if needed (improved).
    - Rest is unchanged.

- 8 bits texture enabled: (improved)
    - On `Source` creation:
        - Object is marked for needing a palette texture and clut content copy (this avoids buffer allocation and clearing on `Source` creation).
    - On `Source` cache lookup:
        - If first lookup then:
            - The `PaletteMap` is queried (i.e. clut hash check plus `compare64`) for a valid `GSTexture`, clut buffer pair and the references are set in the `Source` object using `GSRenderer` clut as reference,
        - else:
            - The copy of the clut referenced by the `Source` object is used to check against the current `GSRenderer` clut, and if mismatch then again the `PaletteMap` is queried for a valid `GSTexture`, clut buffer pair and references are updated.

        Please note that clut buffer copies in CPU, that before were always performed, are done only on `PaletteMap` miss.
        Also, CPU to GPU palette upload is done only on miss.
    - If palette associated to the texture changes:
        - As said above, the `PaletteMap` is in charge of either finding or creating a new `GSTexture`, clut buffer pair.
    - On `Source` deletion:
        - Do nothing.


The new management shows small to none performances improvement in almost every game in terms of FPS, and it lowers the GPU BUS usage by some percentage points in games like Baldur's Gate: Dark Alliance (9% to 7%) which uses many palettes.

The hot topic is that the performances in Zone Of The Enders 2 *skyrocketed*, because of the fact that the games uses many palettes and a small number of textures to render it's effects.
I measured the `PaletteMap` size ~450 palettes in the most concitated moments, which has almost no impact on CPU and GPU memory usage.
I'd like to leave two screenshots for performances comparison, of the game running on my machine (i5-2500k@4.3GHz, GTX770 OC, 16GB DDR3-1600, Win7 64bit), with GSdx-SSE4 32 bit D3D11, default config and 8 bit textures enabled.

Here is the before:

![zoe2](https://user-images.githubusercontent.com/15114320/37565339-d6f4edc6-2aa7-11e8-87ee-040d12132e0f.PNG)

Here is the after:

![zoe2_repo](https://user-images.githubusercontent.com/15114320/37565341-d98c606e-2aa7-11e8-8a9a-86b63965190d.PNG)

More details are coming as edit of this PR in the next days.

TODO: Check interactions with ICO, as the `GSRendererOGL` allocates a palette for that game outside the `GSTextureCache` scope, thus clashes with the new `PaletteMap`.

lightningterror commented:
That's almost double! You should have removed the framelimiter to see the max fps it can reach :stuck_out_tongue:

gregory38 commented:
Cool. I'm not surprised of the perf impact. The idea was for Zone of Enders. But maybe some games use a similar mega-texture pattern.
The really sad news is that we have yet another game that doesn't need Vulkan to be fast !

lightningterror commented:
Tell that to the people that keep whining with_Vulkan when ?_ xD

lightningterror commented on plugins/GSdx/GSTextureCache.h:
```diff
@@ -105,6 +105,47 @@ class GSTextureCache
+		std::array<Palette, MAX_SIZE> m_palettes_stack; // Stack of Palette to be used to avoid memory fragmentation
+		
+		// Array of 2 multimaps, the first for 64B palettes and the second for 1024B palettes.
+		//     Each multimap has the clut hash value as key and the relevant palettes as values (multiple values possible for each key)
```
Small nitpick, extra spaces here.

AlessandroVetere replied:
Oh I put those by purpose, but yes maybe it's better without them!

gregory38 commented on plugins/GSdx/GSTextureCache.cpp:
```diff
@@ -2043,3 +2033,119 @@ void GSTextureCache::SourceMap::RemoveAt(Source* s)
+	// No Palette with matching clut content hash, MISS
+
+	// Create new Palette
+	m_palettes_stack[m_size] = Palette(renderer, pal);
```
Is the palette stack really useful ?

AlessandroVetere replied:
Well, I was experimenting with different things, I don't really know the performance impact of the stack, but for small objects it may be good (also increase probability of false HIT in the hash map being a cache hit).

gregory38 replied:
I understand. I think the cache is already complex enough ;) Could you test the performance without this intermediate structure.

willkuer commented:
Is GS still the bottleneck for ZoE or is it now the EE? What do the indicators show?

atomic83GitHub commented:
This PR cause a regression in Jak 2, I have random crash with this error:

![capture](https://user-images.githubusercontent.com/26351275/37615823-79638e80-2bae-11e8-93d9-d3cec46edb5c.PNG)

lightningterror commented:
Maybe also include a block dump(+savestate) so Alessandro can track down the source faster.

gregory38 commented:
Code is limited to 2048 palettes which aren't enough. Tbh, I think we should remove the intermediate stack containers and multiply palette by 10 or 100. We might need to purge all palette from time to time.

AlessandroVetere commented on plugins/GSdx/GSRendererOGL.cpp:
```diff
@@ -1344,12 +1344,16 @@ void GSRendererOGL::DrawPrims(GSTexture* rt, GSTexture* ds, GSTextureCache::Sour
 
 			// We need the palette to convert the depth to the correct alpha value.
 			if (!tex->m_palette) {
```
Well, is the code is executed also with 32 bit textures? I didn't really inspected well the code, but to me seemed so. It may be needed a bit of refactoring in any case.

gregory38 replied:
On GPU it is a depth texture (so float number) but on GS side it is a PSM_PSMT8H (i.e. 8 bits palettes + 24 bits padding).

I don't remember why I did this ! Maybe we could just always load the clut from the texture cache (even when format is special). It should be mostly free now. I really should post my gsdump somewhere.

AlessandroVetere commented:
@willkuer I did not put attention to the indicators in fact I was only looking at CPU and GPU usage vs FPS.
In fact OpenGL performances were not increased at all (25 fps before and after) so there are still bottlenecks (GS vs EE I will inspect) for this specific game.

AlessandroVetere commented:
@atomic83GitHub Well, if this can be solved by increasing the max number of palettes stored in the map, as @gregory38 suggested, it's no problem, but if strange effects (like shadows or flare) are emulated with clut maybe the number of palettes may be too big.
Also purging palettes requires a bit of refactoring due to the different scopes when destructing Source objects and such to keep track of the number of referencing Source(s) for each palette, but it is doable indeed.

gregory38 commented:
Opengl speed is strange. Did you try to change blending accuracy setting ?

gregory38 commented:
Some palette are used as coefficient factor to modulate intensity. So yes you could have little variation on palette.
It depends on how many palette you could handle in the multimap. (If hash is good enough, btw sse?). Palette is 1kB.  100K will give you 100MB of memory, it should be doable even for old PC.

gregory38 commented on plugins/GSdx/GSTextureCache.cpp:
```diff
@@ -2043,3 +2033,119 @@ void GSTextureCache::SourceMap::RemoveAt(Source* s)
 }
+
+// Query the PaletteMap for a valid Palette, then assign both palette texture pointer and clut copy pointer to the Source object
+void GSTextureCache::SetPalette(Source* s, const GSRenderer* renderer, uint16 pal)
```
Maybe name it `AttachPaletteToSource` or `PaletteLookup`.

gregory38 commented on plugins/GSdx/GSTextureCache.cpp:
```diff
@@ -1598,9 +1585,12 @@ GSTextureCache::Source::Source(GSRenderer* r, const GIFRegTEX0& TEX0, const GIFR
-	m_renderer->m_dev->Recycle(m_palette);
-
-	_aligned_free(m_clut);
+	if (!m_should_have_tex_palette) {
```
I don't like this solution. Desctruction should be done in Palette object. I don't know what will be the speed impact but maybe you could use `shared_pointer<T>` instead. That being said, we should delete palette only when there are unused from a long time (potentially palette will be reused next frame)

AlessandroVetere replied:
The code path which recycles palette on Source destruction is potentially hit only by ICO (palette allocated outside the `GSTextureCache` scope when 32 bit textures, so `m_should_have_tex_palette` is false), but if this is not the case when the `PaletteMap` is not used, then `m_palette` in Source is always null and no recycling outside `Palette` destruction happens.
Desctruction is already done in `PaletteMap` in fact for all the normal cases, and only on `Clear`, which is called when `RemoveAll` is called in the `GSTextureCache`.

gregory38 replied:
It feels way too complex for me. ICO should be updated to a GetPallette (or SetPalette). It should work as the texture cache.

Potentially it could be interesting to use some kind of shared_ptr to handle the reference counting of allocation.

lightningterror replied:
Can a better detection for ICO be added, one that doesn't rely on crc lookup ?

gregory38 commented on plugins/GSdx/GSTextureCache.cpp:
```diff
@@ -2043,3 +2033,119 @@ void GSTextureCache::SourceMap::RemoveAt(Source* s)
+}
+
+// Retrieves the palette with the desired clut
+GSTextureCache::Palette* GSTextureCache::PaletteMap::GetPalette(const GSRenderer* renderer, uint16 pal) {
```
I will call it `LookupPalette`, just to highlight that it is more complex than just get a variable.

AlessandroVetere commented:
@atomic83GitHub If you could please share a dump I can better profile the game behavior with the new code, thanks.
If so many palette are used it is maybe  to emulate intensity variation so it is a good benchmark for the `PaletteMap` scaling capabilities and potential performance improvement.

gregory38 commented on plugins/GSdx/GSTextureCache.cpp:
```diff
@@ -2043,3 +2033,119 @@ void GSTextureCache::SourceMap::RemoveAt(Source* s)
+
+// Hashes the content of the clut
+size_t GSTextureCache::PaletteMap::HashClut(uint16 pal, const uint32* clut) {
+	size_t clut_hash = 3831179159;
```
Add some comments to explain where you did get those hash value.

AlessandroVetere replied:
Big prime number out of the magic hat 🎩 
Jokes aside, I tried to build a simple hash function and this seemed to work well.
It can be improved by hashing 16 indices at a time, using different prime numbers (they have magic properties when hashing) for each index.
The hash function is critical and open to improvements.
I didn't find any SSE method which is "easy" to implement for hashing.
The use case is very specialized though, so maybe hashing 16 values (1 iteration for pal == 16, 16 iterations for pal == 256) at a time with different combinations of operations/prime numbers may give good results.

gregory38 replied:
Comment will be nice. This way, people can test various hash function to see the impact.

lightningterror replied:
Also I prefer if void and similar functions being in a new line like this
```
void something
{
}
```
instead of `void something (`

atomic83GitHub commented:
Here is the Jak 2 dump.

https://drive.google.com/open?id=1zLrGZHyBBhrwv1Djc59GpOP1e5Gus43F

AlessandroVetere commented:
I'm sorry I can't replay the dump, rundll32 crashes when reading the file.
I extracted the .7z and tried to run the .dump file in the folder, should I do something else?

MrCK1 commented:
Oh he gave you a blockdump for diagnosing core issues. I assume you were requesting a GS dump instead?

AlessandroVetere commented:
Yes I was requesting a GS Dump, my bad for being unclear, sorry.

atomic83GitHub commented:
@AlessandroVetere don't worry, Here is the GSDUMP :).

https://drive.google.com/open?id=11uI3f1eRJ7OMq8UikfPO8odU2DRAxpVD

lightningterror commented:
Hint: in the future for single frame block dumps you can just upload them on github. Just rename the file from .xz to .zip :)
They shouldn't be that large and can be done I guess most of the time.

gregory38 commented:
Before I forget it again. OpenGL is quite slow to update the uniform buffer. ZoE rendering will ping-pong between multiple clamping region. Potentially it is the main reason why it is slow. (The other possibility is the multiple blending state change)

gregory38 commented:
What is the current status ?

On thing to do
* Remove the stack (except if it give real perf gain)
* Improve the ICO hack.

AlessandroVetere commented:
The stack of `Palette` is removed without any noticeable performance loss, with increased code maintainability.

The maximum size of the data structure is now 65535 elements, but it can be further increased (it is enforced manually and governs the initial capacity of the two multimaps containing <hash, Palette*> couples).

Some methods have been renamed and a comment have been reformatted.

Regarding ICO I didn't test the game, but I am interested to know if the asserts that I added to the specific code in `GSSource::~Source()` are ever triggered: if they aren't, the code can be removed (it means that the specific code path for ICO in `GSRendererOGL` is never hit when 8 bit texture is enabled).

I also added a reference counter to the Palette object, and it could be used to clear the PaletteMap of unused objects whenever needed (on maximum size reached?), which currently is not done yet.

gregory38 commented:
Thanks you. I will try to review/understand the code and I post you a gsdump for ICO in coming days.

gregory38 commented:
Try this gsdump for ico. Based on the name it should be the good one (I hope!). Uncompress the zip and replay the .xz file directly
[ico.zip](https://github.com/PCSX2/pcsx2/files/1964865/ico.zip)

AlessandroVetere commented:
The GSDump in question made clear that when 8-bit texture is enabled ICO does not allocate a new palette texture in `GSRendererOGL`, so the specific code path added for that corner case is removed!

The pushed commits also include minor updates and fixing.

Still TODO (maybe) is designing a policy to clear unused cached palettes.
And doing some other benchmarks!

AlessandroVetere commented:
@willkuer As for the ZoE 2 EE and GS indicators, I get EE ~93% and GS ~98% (averages) with 8-bit texture enabled and DX11-HW.
In OGL-HW instead EE ~35% and GS ~25%. But maybe that can be discussed in a specific issue and addressed in a different PR.

gregory38 commented on plugins/GSdx/GSTextureCache.h:
```diff
@@ -105,6 +106,55 @@ class GSTextureCache
+		__forceinline GSTexture* GetPaletteGSTexture();
+
+		// Getter for the reference counter
+		__forceinline uint16 GetRefCounter();
```
It feels like it would be better to use `std::shared_ptr<Palette>` pointer to manage the reference counting.

AlessandroVetere replied:
Well the scope of usage is limited, for now. I don't feel confident with `std::shared_ptr` yet, but it may help avoid some extra lines of code.

gregory38 commented on plugins/GSdx/GSTextureCache.cpp:
```diff
@@ -1449,19 +1439,21 @@ GSTextureCache::Source* GSTextureCache::CreateSource(const GIFRegTEX0& TEX0, con
 			src->m_texture = m_renderer->m_dev->CreateTexture(tw, th);
+
+			if (psm.pal > 0) {
+				uint16 palette_size = psm.pal * sizeof(uint32);
```
This parts of code bugs me. Because it is mostly a copy-past of what it is done when you create/attach a palette.

If I'm right this code is for texture that technically need a palette but texture conversion was done on CPU. 
Do you always need a storage for the CPU side ? In this case, it would be more logical to do in the source constructor. Or did I misunderstand something ?

AlessandroVetere replied:
I left the code path for the texture that needs clut copy but have texture conversion done by CPU unchanged.
In fact the clut copy is necessary to ensure correct hit when calling `LookupSource`. I don't think caching clut copies would enhance performances as there is no CPU-GPU communication, and the memcpy alone used here is probably as fast as an hash table hit for clut array.
Also IMHO it is better to move the copy of the clut for CPU converted textures in Source constructor, as the deletion of the array for those is done on Source destruction.
This is therefore out of the scope of the new caching system but for sure I look forward to implement it.

Helium-4 commented:
I gave a couple of the builds a go, the difference is pretty sizable in most of ZoE2 even if it's still isn't enough for full speed everywhere on my PC (4770k@4.2Ghz, GTX 1070), the dips are practically limited to a couple spots in the game and very specific times in those spots (mostly just the Anubis fights and only if you get close to him or if you decide to fire a flurry of homing laser straight into an ice wall in the beginning)

More interesting however is that MGS3 is also affected albeit in a lesser way
From what I've tested in a single spot (5 runs of that test area lowest and highest for all runs combined)
Latest master (v1.5.0-dev-2371-gfbc0b1849)
OGL: 74-82fps
DX11: 78-86fps
The latest available build from this PR as of this post (pcsx2-v1.5.0-dev-2386-g2473e19fd)
OGL: 79-90fps
DX11: 92-99fps

lightningterror commented:
What's the status with Jak 2 ? Was the regression fixed ?

atomic83GitHub commented:
I don't anymore have the error but now, my entire Windows OS sometimes froze or PCSX2 shutdown itself

lightningterror commented:
@AlessandroVetere what's the status on the PR ?

AlessandroVetere commented:
@lightningterror I still need to implement the purge of unused palettes when the maps reach a target max size, which is an easy task and can be done in the same method where new palettes are created.
In my opinion the detection of which palettes are unused on map purge can be done using the ref counter implemented ad hoc for the case, as it seems to me being the simplest solution.
If instead shared pointer usage is required, additional refactoring is needed.
Also there are still some comments to be added and some code needs to be refactored.

@atomic83GitHub Do you experience any performance difference in Jak with this PR? Given the high number of palettes used by the game (which made the previous version of the PR crash) it might be the case, and it might be nice.

gregory38 commented:
Shared_ptr are designed to handle this kinds of shared pointer in multiple structure. They are part of the standard and should be easy to use (even easier and safer than manual ref counting).


Jak renders directly to palette. I suspect they used it to apply  some lighting corrections. So there are likely small variations in the palette. It could explain the huge gain.

By the way, If I were you, I would do something like that
```
if (map is too big)
    Map.Erase_if(lambda [ ref counter is 1 ])
```
So you keep all active entries alive

atomic83GitHub commented:
@AlessandroVetere The speed gain goes from 1/2% to 20/25%. it really depends on where you are, for exemple, the "Under port" level is really improve in terms of performance.

![capture](https://user-images.githubusercontent.com/26351275/40545306-aa14a170-602b-11e8-82e9-9869e88524fc.PNG)

lightningterror commented:
How's progress going ? It's been a while.

AlessandroVetere commented:
I know that I've been M.I.A. for several months (due to work, other projects, 
personal commitments etc...), but here it is my last contribution.

I implemented the logic to clear the `multimap` of unused (detection done via custom reference counter, which counts how many `Source` objects are keeping a reference to the current `Palette`) objects if its size becomes too big, before inserting a new element.
In this way we should keep the map relatively clean as the game progresses.

At this point, I think that after some more testing, a rebase of the PR could be done and, if everything is ok, it could be merged.

AlessandroVetere commented on plugins/GSdx/GSTextureCache.h:
```diff
@@ -59,8 +59,9 @@ class GSTextureCache
 		void Flush(uint32 count, int layer);
 
 	public:
+		void* m_palette_obj; // Pointer to the relevant Palette object (if any)
```
The problem is that the definition of `Palette` happens later in the code, with respect to that member definition, so it cannot be referenced yet, but I can pre-define an empty Palette class to avoid compilation error, though.

turtleli replied:
I think it's also possible to rearrange the order so you don't have to forward declare Palette.

AlessandroVetere commented on plugins/GSdx/GSTextureCache.h:
```diff
@@ -105,6 +106,58 @@ class GSTextureCache
+		
+		// Array of 2 multimaps, the first for 64B palettes and the second for 1024B palettes.
+		// Each multimap has the clut hash value as key and the relevant palettes as values (multiple values possible for each key)
+		std::array<std::unordered_multimap<size_t, Palette*>, 2> m_multimaps;
```
Well, keys are of type `size_t` (the result of the clut hashing function), so for that type are defined both equality and ordering. Using `multimap` would slow down inserts as the structure keeps data ordered by keys, and I don't need that functionality in the code, so `unordered_multimap` seems betters to me here.

turtleli replied:
Not sure if you misread. I wrote `unordered_map`, not `multimap`.

AlessandroVetere replied:
Sorry I misread. In that case the answer is that the key we store for each clut is the hash of its content, in `size_t` format and given the non perfection of the hashing function, different cluts may have the same hash value. The `unordered_map` works if for each key you have a single value, but in our case for each key (hash of the clut) we may have multiple cluts. So either we use `unordered_map<size_t, list<Palette*>>` or we use directly the `unordered_multimap<size_t, Palette*>`, which supports directly multiple values for the same key. Does this answer the question?

turtleli replied:
I mean something similar to:
```
struct PaletteKey {
    const uint32 *clut;
    uint16 pal;
};

struct PaletteKeyHash {
    std::size_t operator()(const PaletteKey &key) const
    {
        // Calculate
        return hash;
    }
};

struct PaletteKeyEqual {
    bool operator()(const PalletteKey &lhs, const PalletteKey &rhs) const
    {
        // Compare clut contents
        return result;
    }
};

std::unordered_map<PaletteKey, Palette*, PaletteHash, PaletteKeyEqual> map;
```
could work.

AlessandroVetere commented on plugins/GSdx/GSTextureCache.cpp:
```diff
@@ -2043,3 +2035,183 @@ void GSTextureCache::SourceMap::RemoveAt(Source* s)
+			palette->Dispose(renderer); // Dispose referenced palette object
+		}
+		multimap.clear(); // Clear all the nodes of the multimap
+		multimap.reserve(MAX_SIZE); // Ensure multimap capacity is not modified by the clearing
```
I cannot find any reference to that, I can remove the reserve but I'm genuinely curious.

turtleli replied:
It's kinda an assumption. It's more efficient to reuse the memory than to free everything and then have to reallocate memory if the container is refilled.

First answer at https://stackoverflow.com/questions/39313820/stdunordered-map-does-not-release-memory kinda shows this behaviour for VS2015. I'd expect most other implementations avoid aggressively freeing memory when its not strictly necessary.

AlessandroVetere commented on plugins/GSdx/GSTextureCache.h:
```diff
@@ -105,6 +106,58 @@ class GSTextureCache
+		__forceinline void DecrementRefCounter();
+
+		// Recycles palette texture and frees clut copy
+		__forceinline void Dispose(const GSRenderer* renderer);
```
It was just to avoid keeping the reference to the renderer saved in each object, but I agree with the fact that it would enable doing the cleanup in the destructor, which is really nicer. Will refactor.

AlessandroVetere commented on plugins/GSdx/GSTextureCache.cpp:
```diff
@@ -2043,3 +2035,183 @@ void GSTextureCache::SourceMap::RemoveAt(Source* s)
+	}
+
+	// Create new Palette
+	Palette* palette = new Palette(renderer, pal);
```
Right, this has been fixed, but +1 for the usage of shared pointer.

AlessandroVetere commented on plugins/GSdx/GSTextureCache.h:
```diff
@@ -105,6 +106,58 @@ class GSTextureCache
+		Palette(const GSRenderer* renderer, uint16 pal); // Creates a copy of the current clut and a texture with its content
+
+		// Getter for clut pointer
+		__forceinline uint32* GetClut();
```
I removed them and I can't distinguish the performance benefit in using `__forceinline` from noise. They were mainly put to allow for "free calls" to getters / small functions. They are now gone.

lightningterror commented:
@AlessandroVetere can you update the PR with the changes requested by turtleli if they are better, also can you also do a rebase, I believe this PR is really close for a merge.

lightningterror commented:
@gregory38 how does it look ?

AlessandroVetere commented:
Just a brief of description of the PR status.

Finally used `std::shared_ptr<Palette>` instead of `Palette*` plus manual object lifecycle management.
Also moved `class Palette` declaration up in the source file to avoid forward declaration.
I kept the `reserve` after `clear` of `unordered_multimap`.
I plan to add some comments to the clut hashing function, to allow for better understanding of the current choice and help improving it in the future (Bitwise only operations? SSE / AVX codes?); for now, the current hashing function just does the work, with very small percentage of hash collisions (evaluated on the small test cases I have available).

I would skip implementing the custom `unordered_map` as it would need a bit of refactor of the current code, which uses the more complex `unordered_multimap` container but in its "standard" form.

Are there any other suggestion/comment/things to implement (comments to hashing function aside)?

turtleli commented:
> I would skip implementing the custom unordered_map as it would need a bit of refactor of the current code, which uses the more complex unordered_multimap container but in its "standard" form.

It feels like you are kind of implementing unordered_map with unordered_multimap. A hash lookup is done to find the bucket (equal_range()) and then the first exact clut match is used because there won't be duplicate entries.

It's up to you whether you change it, but IMO the "custom" unordered_map would make it clearer that there is one palette per clut.

lightningterror commented on plugins/GSdx/GSTextureCache.cpp:
```diff
@@ -2043,3 +2034,164 @@ void GSTextureCache::SourceMap::RemoveAt(Source* s)
+	GL_CACHE("TC, %u-bit PaletteMap (Size %u): Added new palette.", palette_size, multimap.size());
+	
+	// Return the shared pointer to the newly created Palette
+	return palette;
```
I think it's fine, it makes the code easier to understand especially for new people.

turtleli replied:
If the comments simply restate, line for line, what the code is doing then it doesn't actually help the reader understand what the intent of the code is.

ramapcsx2 replied:
Here the correctness of the implementation is very important. 
These verbose comments can help ensure that the code actually does what the comments say it should do.
So I agree, these comments are helpful here.
(Of course, not so much if this were some UI code ;p)

gregory38 replied:
Honestly I agree with turtleli here. Those comments don't add any relevant information.

Syst3mSh0ck replied:
Consistently well documented code should be welcomed and encouraged :)

ramapcsx2 commented on plugins/GSdx/GSTextureCache.cpp:
```diff
@@ -2043,3 +2034,164 @@ void GSTextureCache::SourceMap::RemoveAt(Source* s)
+// Copy constructor
+GSTextureCache::Palette::Palette(const Palette& palette) {
+	m_clut = palette.m_clut;
+	m_tex_palette = palette.m_tex_palette;
```
@turtleli Do you have a recommendation what to do here? A deep copy is probably not what was intended.
Do double free's always crash or can we fix it with a guard of sorts?
Thanks!

turtleli replied:
Deleting the copy constructor and copy assignment operator would be a good idea if they're not meant to be copied (I think they're not currently used).
```
Palette(const Palette&) = delete;
Palette& operator=(const Palette&) = delete;
```

It'll trigger a compile error if someone unknowingly tries to copy it.

The move constructor and move assignment operators could be deleted too if they're not meant to be moved.

AlessandroVetere commented on plugins/GSdx/GSTextureCache.cpp:
```diff
@@ -62,6 +62,8 @@ GSTextureCache::GSTextureCache(GSRenderer* r)
 	// Test: onimusha 3 PAL 60Hz
 	m_temp = (uint8*)_aligned_malloc(9 * 1024 * 1024, 32);
+
+	m_palette_map.SetRenderer((const GSRenderer*) r);
```
The reference to the `GSRenderer` in `GSTextureCache` is not `const`, so have to cast it to `const`. However I can drop the specifier in the members of the new classes and remove the cast there if it seems unneded.

turtleli replied:
It's usually not necessary to explicitly cast a non-const pointer to const.

AlessandroVetere commented on plugins/GSdx/GSTextureCache.cpp:
```diff
@@ -2043,3 +2034,164 @@ void GSTextureCache::SourceMap::RemoveAt(Source* s)
+	// No Palette with matching clut content hash, MISS
+
+	if (multimap.size() > MAX_SIZE) {
+		// If the multimap is too big, try to clean it by disposing and removing unused palettes, before adding the new one
```
If I remove the cleaning logic performed if the size of the map exceeds a given threshold, the size of the map would only increase with time, as the palettes remains in memory independently of the lifecycle of its users (the `Source` objects).
The expected result of this cleaning logic is that many palettes in the map gets cleared in a single shot, because the number of `Source` objects that can be active on screen at the same time is much smaller than `MAX_SIZE` leaving many palettes unused at each frame (but can be reused at the next, maybe), and thus at the next lookup call, if the result is miss + insert new palette, the cleaning logic is not triggered as the map was reduced greatly in size.
In the tests I performed, the logic was rarely hit, but the maximum size could be tuned to have a more aggressive cleaning of the unused palettes, in case the lookup slows down too much the rendering (which I didn't experience, though); or vice versa.

ramapcsx2 commented:
Half a year in the making, is it time for merging this yet? ;)

lightningterror commented:
@AlessandroVetere are you ok if we finish up the PR and merge it ? @tadanokojin has the necessary changes that are requested ready.

AlessandroVetere commented:
I actually pushed some commits fixing/improving what was left in my fork, aside of the suggested usage of `unordered_map` instead of `unordered_multimap` as base structure for caching the palettes.
That last change would require careful refactoring, which I think could be done in a different PR after this has been merged, if the necessity occurs.

I think that now, after a little testing of the build, the commits of the PR could be merged with squashing, if everybody agrees.

lightningterror commented:
Sure, I don't mind if it's done in a separate PR, are you gonna squash the commits or do them in merge-squash ?

AlessandroVetere commented:
I'm on my way to deliver also the `unordered_map` container for the palette cache...it's an easier task than expected!

lightningterror commented:
Nice, I'll merge this once it's added.

AlessandroVetere commented:
The change has been pushed!
Thanks @turtleli for the `unordered_map` hint, the caching solution is much clearer this way.

lightningterror commented:
Travis build seems to fail with the error
```
/home/travis/build/PCSX2/pcsx2/plugins/GSdx/GSTextureCache.cpp:2190:75: error: ‘palette_size’ was not declared in this scope
    GL_INS("INFO, %u-bit PaletteMap (Size %u): Cleared %u palettes.", palette_size, map.size(), cleared_palette_count);
```

lightningterror commented:
There's also some warnings as well
https://travis-ci.org/PCSX2/pcsx2/jobs/450450991

AlessandroVetere commented:
My bad, VS was not saying a thing about the missing variable declaration! And as for the warnings I'll fix those as well, in a couple of hours I can get to PC and finish the work.

AlessandroVetere commented:
Errors and warnings fixed! :tada:

lightningterror commented:
Great work, and merged it. Looking forward to your next project :)
